servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,681 @@
+                    final int prevActiveMappedSources = activeMappedSources;
+                    if (prevActiveMappedSources < 0) {
+                        // We have been cancelled, so in this case we don't want to Subscribe and instead
+                        // just want to discard the Publisher.
+                        subscribers.remove(subscriber);
+                        break;
+                    } else if (activeMappedSourcesUpdater.compareAndSet(this, prevActiveMappedSources,
+                            prevActiveMappedSources + 1)) {
+                        publisher.subscribeInternal(subscriber);
+                        tryRequestMore(subscriber);
+                        break;
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            if (!onError0(t, false, false)) {
+                LOGGER.debug("Already terminated/cancelled, ignoring error notification.", t);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            // Setting terminal must be done before terminateActiveMappedSources to ensure visibility of the terminal.
+            final boolean setTerminal = trySetTerminal(complete(), false, terminalNotificationUpdater, this);
+            final boolean allSourcesTerminated = terminateActiveMappedSources();
+            if (setTerminal && allSourcesTerminated) {
+                enqueueAndDrain(complete());
+            }
+        }
+
+        private boolean onError0(Throwable throwable, boolean overrideComplete,
+                                 boolean cancelSubscriberIfNecessary) {
+            final TerminalNotification notification = TerminalNotification.error(throwable);
+            if (trySetTerminal(notification, overrideComplete, terminalNotificationUpdater, this)) {
+                try {
+                    doCancel(cancelSubscriberIfNecessary);
+                } finally {
+                    enqueueAndDrain(notification);
+                }
+                return true;
+            }
+            return false;
+        }
+
+        private void doCancel(boolean cancelSubscription) {
+            Throwable delayedCause = null;
+            // Prevent future onNext operations from adding to subscribers which otherwise may result in
+            // not cancelling mapped Subscriptions.
+            activeMappedSources = Integer.MIN_VALUE;
+            try {
+                if (cancelSubscription) {
+                    assert subscription != null;
+                    subscription.cancel();
+                }
+            } finally {
+                for (FlatMapPublisherSubscriber<T, R> subscriber : subscribers) {
+                    try {
+                        subscriber.cancelFromUpstream();
+                    } catch (Throwable c) {
+                        if (delayedCause == null) {
+                            delayedCause = c;
+                        }
```

Comment:
should we use `addSuppressed(...)` if `delayedCause != null` ?

Reply:
throwing from methods in this scope isn't supported, so re-throwing the first cause is a compromise to provide context in the event an exception does happen but doesn't attempting to retain all exceptions that occur along the way. `addSuppressed` can also be painful for static/cached exceptions.

Reply:
Fair enough... that said just because you not use `addSuppressed` here doesn't mean that a user will not do it later on ;) So in any case special care need to be taken for static / cached exceptions